server/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java

```diff
@@ -401,48 +402,42 @@ Set<String> matchingNodeIds(DiscoveryNodes nodes) {
-            )
+        List<VersionInformation> dataVersions = List.of(
+            new VersionInformation(Version.fromString("3.2.5"), IndexVersion.fromId(2000099), IndexVersion.fromId(3020599)),
+            new VersionInformation(Version.fromString("3.0.7"), IndexVersion.fromId(2000099), IndexVersion.fromId(3000799)),
+            new VersionInformation(Version.fromString("2.1.0"), IndexVersion.fromId(1000099), IndexVersion.fromId(2010099))
         );
-        discoBuilder.add(
-            new DiscoveryNode(
-                "name_" + 2,
-                "node_" + 2,
-                buildNewFakeTransportAddress(),
-                Collections.emptyMap(),
-                new HashSet<>(randomSubsetOf(DiscoveryNodeRole.roles())),
-                Version.fromString("6.3.0")
-            )
+        List<VersionInformation> observerVersions = List.of(
+            new VersionInformation(Version.fromString("5.0.17"), IndexVersion.fromId(0), IndexVersion.fromId(5001799)),
+            new VersionInformation(Version.fromString("2.0.1"), IndexVersion.fromId(1000099), IndexVersion.fromId(2000199)),
+            new VersionInformation(Version.fromString("1.6.0"), IndexVersion.fromId(0), IndexVersion.fromId(1060099))
         );
-        discoBuilder.localNodeId("node_" + between(1, 3));
-        if (randomBoolean()) {
-            discoBuilder.masterNodeId("node_" + between(1, 3));
+
+        DiscoveryNodes.Builder discoBuilder = DiscoveryNodes.builder();
+        for (int i = 0; i < dataVersions.size(); i++) {
+            discoBuilder.add(
+                DiscoveryNodeUtils.builder("data_" + i)
+                    .version(dataVersions.get(i))
+                    .roles(Set.of(randomBoolean() ? DiscoveryNodeRole.DATA_ROLE : DiscoveryNodeRole.MASTER_ROLE))
+                    .build()
+            );
+        }
+        for (int i = 0; i < observerVersions.size(); i++) {
+            discoBuilder.add(DiscoveryNodeUtils.builder("observer_" + i).version(observerVersions.get(i)).roles(Set.of()).build());
         }
-        discoBuilder.add(
-            new DiscoveryNode(
-                "name_" + 3,
-                "node_" + 3,
-                buildNewFakeTransportAddress(),
-                Collections.emptyMap(),
-                new HashSet<>(randomSubsetOf(DiscoveryNodeRole.roles())),
-                Version.fromString("1.1.0")
-            )
-        );
         DiscoveryNodes build = discoBuilder.build();
-        assertEquals(Version.fromString("6.3.0"), build.getMaxNodeVersion());
-        assertEquals(Version.fromString("1.1.0"), build.getMinNodeVersion());
+
+        assertEquals(Version.fromString("5.0.17"), build.getMaxNodeVersion());
+        assertEquals(Version.fromString("1.6.0"), build.getMinNodeVersion());
+        assertEquals(Version.fromString("2.1.0"), build.getSmallestNonClientNodeVersion());
+        assertEquals(IndexVersion.fromId(2010099), build.getMaxDataNodeCompatibleIndexVersion());
```

Comment:
This ID loosely translates to "2.0.0", right? Why is it the max data node compatible index version? Is it basically a way of saying "2.x"?

Reply:
This is the lowest max index version of all nodes that can store data